File: src/SimpleVAO.cpp
#include "SimpleVAO.h"
namespace ngl
{
SimpleVAO::~SimpleVAO()
{
  removeVAO();
}

void SimpleVAO::draw() const
{
  if(!m_allocated)
  {
    NGLMessage::addWarning("Trying to draw an unallocated VOA");
  }
  if(!m_bound)
  {
    NGLMessage::addWarning("Warning trying to draw an unbound VOA");
  }
  glDrawArrays(m_mode, 0, static_cast< GLsizei >(m_indicesCount));
}

void SimpleVAO::removeVAO()
{
  if(m_bound)
  {
    unbind();
  }
  if(m_allocated)
  {
    glDeleteBuffers(1, &m_buffer);
  }
  glDeleteVertexArrays(1, &m_id);
  m_allocated = false;
}

#ifdef PYTHONBUILD

void SimpleVAO::setData(size_t _size, const std::vector< float > &_data)
{
  std::cout << "Setting data from Python float" << _size << '\n';
  if(m_bound == false)
  {
    NGLMessage::addWarning("trying to set VOA data when unbound");
  }
  if(m_allocated == true)
  {
    glDeleteBuffers(1, &m_buffer);
  }
  glGenBuffers(1, &m_buffer);
  // now we will bind an array buffer to the first one and load the data for the verts
  glBindBuffer(GL_ARRAY_BUFFER, m_buffer);
  glBufferData(GL_ARRAY_BUFFER, static_cast< GLsizeiptr >(_size), &_data[0], GL_STATIC_DRAW);
  m_allocated = true;
}

void SimpleVAO::setData(size_t _size, const std::vector< Vec3 > &_data)
{
  std::cout << "Setting data from Python Vec3" << _size << '\n';
  if(m_bound == false)
  {
    NGLMessage::addWarning("trying to set VOA data when unbound");
  }
  if(m_allocated == true)
  {
    glDeleteBuffers(1, &m_buffer);
  }
  glGenBuffers(1, &m_buffer);
  // now we will bind an array buffer to the first one and load the data for the verts
  glBindBuffer(GL_ARRAY_BUFFER, m_buffer);
  glBufferData(GL_ARRAY_BUFFER, static_cast< GLsizeiptr >(_size), &_data[0].m_x, GL_STATIC_DRAW);
  m_allocated = true;
}

#endif

void SimpleVAO::setData(const VertexData &_data)
{
  if(!m_bound)
  {
    NGLMessage::addWarning("trying to set VOA data when unbound");
  }
  if(m_allocated)
  {
    glDeleteBuffers(1, &m_buffer);
  }
  glGenBuffers(1, &m_buffer);
  // now we will bind an array buffer to the first one and load the data for the verts
  glBindBuffer(GL_ARRAY_BUFFER, m_buffer);
  glBufferData(GL_ARRAY_BUFFER, static_cast< GLsizeiptr >(_data.m_size), &_data.m_data, _data.m_mode);
  m_allocated = true;
}

Real *SimpleVAO::mapBuffer(unsigned int _index, GLenum _accessMode)
{
  NGL_UNUSED(_index)
  bind();
  glBindBuffer(GL_ARRAY_BUFFER, m_id);
  auto *ptr = static_cast< Real * >(glMapBuffer(GL_ARRAY_BUFFER, _accessMode));
  // modern GL allows this but not on Mac!
  // ptr = static_cast<Real *>(glMapNamedBuffer(m_id, _accessMode));

  return ptr;
}

} // namespace ngl

File: py-ngl/src/ngl/simple_vao.py
import OpenGL.GL as gl

from .abstract_vao import AbstractVAO, VertexData
from .log import logger


class SimpleVAO(AbstractVAO):
    def __init__(self, mode=gl.GL_TRIANGLES):
        super().__init__(mode)
        self.m_buffer = gl.glGenBuffers(1)

    def draw(self):
        if self.m_bound and self.m_allocated:
            gl.glDrawArrays(self.m_mode, 0, self.m_indicesCount)
        else:
            logger.error("SimpleVAO not bound or not allocated")

    def set_data(self, data):
        if not isinstance(data, VertexData):
            logger.error("SimpleVAO: Invalid data type")
            raise TypeError("data must be of type VertexData")
        if not self.m_bound:
            logger.error("SimpleVAO not bound")
            raise RuntimeError("SimpleVAO not bound")
        gl.glBindBuffer(gl.GL_ARRAY_BUFFER, self.m_buffer)
        gl.glBufferData(gl.GL_ARRAY_BUFFER, data.data.nbytes, data.data, data.mode)
        self.m_allocated = True
        self.m_indicesCount = data.size

    def remove_vao(self):
        gl.glDeleteBuffers(1, [self.m_buffer])
        gl.glDeleteVertexArrays(1, [self.m_id])

    def get_buffer_id(self, index=0):
        return self.m_buffer

    def map_buffer(self, index=0, access_mode=gl.GL_READ_WRITE):
        gl.glBindBuffer(gl.GL_ARRAY_BUFFER, self.m_buffer)
        return gl.glMapBuffer(gl.GL_ARRAY_BUFFER, access_mode)
